pkg/policy/types/requirements.go

```
@@ -0,0 +1,264 @@
+	return reqs
+}
+
+func NewExceptRequirement(lbl labels.Label) Requirement {
+	return Requirement{
+		key:      lbl,
+		operator: selection.DoesNotExist,
+	}
+}
+
+func NewEqualsRequirement(lbl labels.Label) Requirement {
+	return Requirement{
+		key:      labels.Label{Key: lbl.Key, Source: lbl.Source},
+		operator: selection.Equals,
+		values:   set.NewSet(lbl.Value),
+	}
+}
+
+func NewRequirement(key string, op selection.Operator, values []string) Requirement {
+	return Requirement{
+		key:      labels.ParseSelectDotLabel(key),
+		operator: op,
+		values:   set.NewSet(values...),
+	}
+}
+
+func (r *Requirement) GetKeyPrefix() *netip.Prefix {
+	return r.key.GetCIDRPrefix()
+}
+
+func (r *Requirement) HasKeySource(source string) bool {
+	return r.key.Source == source
+}
+
+func (r *Requirement) HasValue(value string) bool {
+	return r.values.Has(value)
+}
+
+func RequirementsFromK8s(k8sReqs k8sLbls.Requirements) Requirements {
+	reqs := make(Requirements, 0, len(k8sReqs))
+	for _, req := range k8sReqs {
+		reqs = append(reqs, Requirement{
+			key:      labels.ParseSelectDotLabel(req.Key()),
+			operator: req.Operator(),
+			values:   set.NewSet[string](req.ShallowValues()...),
+		})
+	}
+	return reqs
+}
+
+// Matches returns true if the Requirement matches the input Labels.
+// This is structurally the same as k8sLbls.Requirement.Matches(),
+// but takes LabelArray as an argument instead of a string.
+//
+// MatchesRequirements is defined as a generic function on labels.LabelMatcher rather than taking
+// the interface as the parameter due to the interface parameter causing Go compiler (1.25.2) to
+// escape labels.LabelArray implementing the interface to the heap, causing large memory
+// overhead. When the compiler specializes the generic type to a concrete type (not an interface),
+// it can perform correct escape analysis and avoid unnecessary heap allocations.<
+func MatchesRequirement[T labels.LabelMatcher](logger *slog.Logger, r *Requirement, ls T) bool {
```

I'd suggest this return an error instead of taking a logger; callers can handle the error as they see fit.

Requirements are parsed from k8s label selectors, which are validated. So in practice I don't think this is possible, but the "chain" is long and seems fragile, so-to-speak.

---

Then I would need to decide on each call site what to do with the error, which would only happen if values were not properly validated.

IMO the "right thing" to do would be to panic, but we should never do that in production. The next best thing to do is to log an error that will fail our CI. If we always log the error then we should do it here instead of all the call sites.

We can get rid of passing the logger by using `logging.DefaulSlogLogger` instead. Since this is a "soft panic" and never expected in production, this would be acceptable, IMO.

---

@squeed Added a commit to address this, please take a look!

---

thanks, that makes things cleaner.